docs/proposals/service-catalog.md

```diff
@@ -0,0 +1,123 @@
+
+- reference: "I want to use this entry as-is, including its configuration data"
+- template: "I want to create items from the specified template"
+- provision: "I want the creation to be goverened by some other entity that implements the 'service broker' HTTP interface"
+
+We see the type as an arbitrary `string`; one or more controllers could run to process each claim type, performing whatever logic is appropriate to fulfill the specific type in question.
+
+### Viewing a catalog
+
+Users should be able to list the entries in a service catalog. Users should be able to select an entry and "consume" it. We call this consumption "claiming" a service from a service catalog.
+
+### Claiming a service
+
+When you want to use an entry from the service catalog, you create a `ServiceClaim` that references the desired entry. A controller processes new claims for admission. This determines if the user who created the claim is allowed to consume the entry from the catalog. This decision can be flexible: it could be automated based on policy, or it could support manual intervention and workflow.
+
+Once the claim has been admitted, a controller performs the provisioning process based on the service catalog entry's type.
+
+For a reference claim, a new service is created in the user's namespace. The new service somehow needs to be a CNAME to the source. Additionally, for each configuration resource referenced by the source service, the controller creates a clone of it in the user's namespace.
+
+For a template claim, the template is processed and its items are created in the user's namespace (see [OpenShift templates](https://docs.openshift.org/latest/dev_guide/templates.html) for more details on templates).
+
+For a provision claim, the controller sends requests to the specified "service broker" endpoint to perform whatever provisioning actions the broker is coded to do.
+
+Additional fulfillment types are possible as long as there is a controller that is handling them.
+
+TODO poking holes in cluster firewall for cross-namespace connections
+
+## Linking services
+
+Claiming a service catalog entry only creates resources in the user's namespace. If all you need is a service and its DNS entry, this may be sufficient for your pods to function. But if you need the configuration data injected into your pod, it would be nice to make that easier to do.
+
+We want to add the ability to link a service to a deployable resource such as a Deployment. This could look something like:
+
+    kubectl link postgresql deployment/web
+    service "postgresql" linked to deployemnt "web"
+    configdata "postgresql-options" linked to deployment "web" as a volume
+    secret "postgresql-credentials" linked to deployment "web" as a volume
+
+This command would automatically inject the ConfigData and Secret objects as volumes into the Deployment. This could be flexible as well, allowing you instead to expose these items as environment variables.
```

Comment:
Think of this use case as "can we expose, by convention, useful patterns
for relating services to their pod consumers".  It should *not*
prevent/supplant more complicated or direct consumption.  Instead, it's
about taking our tools (config data, secrets, and DNS / service discovery)
and making them easily accessible.  Like *_SERVICE_HOST

Yes, a user can add downward API, secrets, volumes, etc to a pod template.
But if that action is truly common (it is) we should think about making
that pattern obvious and / or easy to accomplish.  That may be something
generator-y on the cli (kubectl run), or an endpoint on the API that does
the simplest possible thing that works.


On Nov 23, 2015, at 11:55 AM, Andy Goldstein <notifications@github.com>
wrote:

In docs/proposals/service-catalog.md
<https://github.com/kubernetes/kubernetes/pull/17543#discussion_r45628290>:

> +Additional fulfillment types are possible as long as there is a controller that is handling them.
> +
> +TODO poking holes in cluster firewall for cross-namespace connections
> +
> +## Linking services
> +
> +Claiming a service catalog entry only creates resources in the user's namespace. If all you need is a service and its DNS entry, this may be sufficient for your pods to function. But if you need the configuration data injected into your pod, it would be nice to make that easier to do.
> +
> +We want to add the ability to link a service to a deployable resource such as a Deployment. This could look something like:
> +
> +    kubectl link postgresql deployment/web
> +    service "postgresql" linked to deployemnt "web"
> +    configdata "postgresql-options" linked to deployment "web" as a volume
> +    secret "postgresql-credentials" linked to deployment "web" as a volume
> +
> +This command would automatically inject the ConfigData and Secret objects as volumes into the Deployment. This could be flexible as well, allowing you instead to expose these items as environment variables.

I'll push an update with an example. I'm not yet convinced this has the
right semantics, but it's at least a point for discussion.

As far as why it's not part of a Deployment's declaration, it definitely
can be, but also consider this scenario. You have a web app you want to
deploy, and it needs a database. Using a web UI, you browse through a list
of services, looking for a database. You select a PostgreSQL service and,
once consumed, you receive a username and password. You now use the web UI
to create a Deployment. You want an easy way to somehow link your
PostgreSQL service (including the generated username/password) to your
Deployment. Again, I think this is something that will need a lot of
discussion and iteration before we have a good UX, but this is the type of
scenario we're looking to solve.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17543/files#r45628290>.